### PR TITLE
Integrating Buying Passports with ETHPass

### PR DESCRIPTION
**Pull request by dvargasfuertes**
https://app.clarity.so/cabin/work/491

### CODE REVIEW

File: packages/app/pages/api/ethpass.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+              contractAddress: address,
+              tokenId,
+              signature: "",
+              signatureMessage: "",
+              pass: {
+                type: "generic",
+                description,
+              },
+            },
+            { headers: { "X-API-KEY": process.env.ETHPASS_API_KEY || "" } }
```
Comment (dvargasfuertes) [inline packages/app/pages/api/ethpass.ts]: This is the main API call. It's not clear to me where I should get `signature` and `signatureMessage`? But I think the lack of setting these values is causing the API call to fail

cc: @christopherdro

**Review by dvargasfuertes — COMMENTED** [review]
I have an android, will I be able to test this with the Wallet Passes app?

cc @christopherdro

File: packages/app/pages/api/ethpass.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            "https://api.ethpass.xyz.com/api/v1/pass",
+            {
+              contractAddress: address,
+              tokenId,
+              signature,
+              signatureMessage,
+              pass: {
+                type: "generic",
+                description,
+              },
```
Comment (dvargasfuertes) [inline packages/app/pages/api/ethpass.ts]: Okay I think I'm generating the signature and signatureMessage correctly now, and getting `certificate has expired` 500 error. I think this is where we set up a certificate with Apple/EthPass? cc @christopherdro

Comment (christopherdro) [reply packages/app/pages/api/ethpass.ts]: I'm not sure that error is coming from our service. Just checked the logs and I don't see any incoming requests. 
I'll ping you on discord.

File: packages/app/pages/api/ethpass.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            },
+            {
+              headers: {
+                "X-API-KEY": process.env.ETHPASS_API_KEY || "",
+                "x-api-scope": "pass.com.ethpass",
+              },
+            }
+          )
+        )
+        .then((r) => res.status(200).json(r.data))
```
Comment (christopherdro) [inline packages/app/pages/api/ethpass.ts]: This shape of the response payload currently is `{ id, buffer}` where `id` is the unique id of the pass created, and `buffer` is the pass file in a raw format so you can choose how you want to distribute it. How are you envisioning users installing the pass?

Comment (dvargasfuertes) [reply packages/app/pages/api/ethpass.ts]: we will probably upload it to IPFS. Could we also display it as a QR code that can be scanned right in the checkout?

Comment (christopherdro) [reply packages/app/pages/api/ethpass.ts]: Yes, you can create a QR code with the file url. Here's how it's done on our demo.
```
import QRCode from 'qrcode'

const canvas = document.getElementById('canvas');
const fileURL = "https://"

QRCode.toCanvas(canvas, fileURL, function (error) {
  if (error) console.error(error);
});
```

There are a few caveats when dealing with IPFS. We can't specify content header types when trying to download files from public gateways so this has been causing some issues on older iOS devices. I'm considering adding IPFS as a storage option out the box to onboard users quicker alongside a proxy that will take care of setting the content header types properly. 

In the meantime, if you'd like to test uploading the buffer to IPFS you can use the following code.
```
export const uploadIpfs = async (buffer: Buffer) => {
  try {
    const passName = `DemoPass-${Math.random()}.pkpass`;
    const auth = 'Basic ' + Buffer.from(projectId + ':' + projectSecret).toString('base64');
    const client = create({
      host: 'ipfs.infura.io',
      port: 5001,
      protocol: 'https',
      headers: {
        authorization: auth,
      },
    });
    const response = await client.add(buffer, { pin: false });
    return `https://ipfs.infura.io/ipfs/${response.path}?filename=${passName}`;
  } catch (err) {
    console.log('## IPFS Upload error:', err);
    throw err;
  }
};

File: packages/app/pages/api/ethpass.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      const contract = new web3.eth.Contract(
+        getAbiFromJson(passportJson),
+        address
+      );
+      return (contract.methods.symbol() as ContractSendMethod)
+        .call()
+        .then((description) => {
+          const body = {
+            contractAddress: address,
+            chainId: networkIdByName[network],
```
Comment (dvargas92495) [inline packages/app/pages/api/ethpass.ts:24]: @christopherdro Now I'm seeing this:
image

API Response at Thu, 24 Feb 2022 20:15:25 GMT

Comment (dvargas92495) [reply packages/app/pages/api/ethpass.ts:24]: To be clear, this is the error returned from the API call, it returned this error message as an HTML string

Comment (dvargas92495) [reply packages/app/pages/api/ethpass.ts:24]: Ignore this comment, I forgot the URL change

**Comment by dvargasfuertes** [comment]
Got them working!
ethpassfront
ethpassback

File: packages/app/pages/checkout/[network]/[address].tsx [hunk 5]
```diff
@@ -175,105 +195,142 @@ const CheckoutPage = ({
...
+            (receipt.events?.["Purchase"]?.returnValues?.id as string) || "";
+          setLoading(false);
+          setSupply(supply - 1);
+          resolve(tokenId);
+        })
+        .on("error", reject)
+    )
+      .then((tokenId) => {
+        if (generateApplePass && tokenId) {
+          const signatureMessage = `Give Passports permission to generate an Apple Wallet Pass for token ${tokenId}`;
```
Comment (ssaitejar96) [inline packages/app/pages/checkout/[network]/[address].tsx:232]: Not very clear about the API but is there support for anything similar in Android? If yes, did we decide to do only Apple first?

**Review by ssaitejar96 — APPROVED** [review]
One minor question but LGTM!

File: packages/app/pages/checkout/[network]/[address].tsx (hunk 5 above)
Comment (dvargasfuertes) [reply packages/app/pages/checkout/[network]/[address].tsx:232]: There wasn't support for Android when I started the PR, but there is now. Will be adding Android support via Google Pay on the next PR for this task

**Pull request merged**